analyzers/src/SonarAnalyzer.CSharp/Rules/RedundantDeclaration.cs

```diff
@@ -291,66 +268,56 @@ private static void ReportOnRedundantObjectInitializer(SyntaxNodeAnalysisContext
+            if (argumentExpression == null)
             {
                 return;
             }
 
-            var argumentExpression = objectCreation.ArgumentList?.Arguments.FirstOrDefault()?.Expression;
-            if (argumentExpression == null)
+            if (!IsDelegateCreation(objectCreation, context.SemanticModel))
             {
                 return;
             }
 
-            if (IsInDeclarationNotVarNotDelegate(objectCreation, context.SemanticModel) ||
-                IsAssignmentNotDelegate(objectCreation, context.SemanticModel) ||
-                IsReturnValueNotDelegate(objectCreation, context.SemanticModel) ||
-                IsInArgumentAndCanBeChanged(objectCreation, context.SemanticModel,
+            if (IsInDeclarationNotVarNotDelegate(objectCreation.Expression, context.SemanticModel)
+                || IsAssignmentNotDelegate(objectCreation.Expression, context.SemanticModel)
+                || IsReturnValueNotDelegate(objectCreation.Expression, context.SemanticModel)
+                || IsInArgumentAndCanBeChanged(objectCreation, context.SemanticModel,
                     invocation => invocation.ArgumentList.Arguments.Any(a => IsDynamic(a, context.SemanticModel))))
             {
-                ReportIssueOnRedundantObjectCreation(context, objectCreation, "explicit delegate creation", RedundancyType.ExplicitDelegate);
+                ReportIssueOnRedundantObjectCreation(context, objectCreation.Expression, "explicit delegate creation", RedundancyType.ExplicitDelegate);
             }
         }
 
-        private static bool IsDynamic(ArgumentSyntax argument, SemanticModel semanticModel)
-        {
-            return semanticModel.GetTypeInfo(argument.Expression).Type is IDynamicTypeSymbol;
-        }
+        private static bool IsDynamic(ArgumentSyntax argument, SemanticModel semanticModel) =>
+            semanticModel.GetTypeInfo(argument.Expression).Type is IDynamicTypeSymbol;
 
-        private static bool IsInDeclarationNotVarNotDelegate(ObjectCreationExpressionSyntax objectCreation, SemanticModel semanticModel)
+        private static bool IsInDeclarationNotVarNotDelegate(SyntaxNode objectCreation, SemanticModel semanticModel)
         {
             var variableDeclaration = objectCreation.GetSelfOrTopParenthesizedExpression()
                 .Parent?.Parent?.Parent as VariableDeclarationSyntax;
 
             var type = variableDeclaration?.Type;
 
-            if (type == null ||
-                type.IsVar)
+            if (type == null || type.IsVar)
             {
                 return false;
             }
 
             var typeInformation = semanticModel.GetTypeInfo(type).Type;
 
-            return typeInformation != null &&
-                !typeInformation.Is(KnownType.System_Delegate);
+            return typeInformation != null && !typeInformation.Is(KnownType.System_Delegate);
```

Comment:
Coverage: `typeInformation` will be `ErrorType` and will never be null if I'm not mistaken. Let's remove it to bump coverage and add a test case.